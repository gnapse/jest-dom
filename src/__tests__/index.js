import '../extend-expect'
import {plugins} from 'pretty-format'
import {render} from './helpers/test-utils'

expect.addSnapshotSerializer(plugins.ConvertAnsi)

test('.toBeInTheDOM', () => {
  const {queryByTestId} = render(`<span data-testid="count-value">2</span>`)

  expect(queryByTestId('count-value')).toBeInTheDOM()
  expect(queryByTestId('count-value1')).not.toBeInTheDOM()

  // negative test cases wrapped in throwError assertions for coverage.
  expect(() =>
    expect(queryByTestId('count-value')).not.toBeInTheDOM(),
  ).toThrowError()
  expect(() =>
    expect(queryByTestId('count-value1')).toBeInTheDOM(),
  ).toThrowError()
  expect(() =>
    expect({thisIsNot: 'an html element'}).toBeInTheDOM(),
  ).toThrowError()
})

test('.toHaveTextContent', () => {
  const {queryByTestId} = render(`<span data-testid="count-value">2</span>`)

  expect(queryByTestId('count-value')).toHaveTextContent('2')
  expect(queryByTestId('count-value')).toHaveTextContent(/2/)
  expect(queryByTestId('count-value')).not.toHaveTextContent('21')
  expect(() =>
    expect(queryByTestId('count-value2')).toHaveTextContent('2'),
  ).toThrowError()

  expect(() =>
    expect(queryByTestId('count-value')).toHaveTextContent('3'),
  ).toThrowError()
  expect(() =>
    expect(queryByTestId('count-value')).not.toHaveTextContent('2'),
  ).toThrowError()
})

test('.toHaveAttribute', () => {
  const {queryByTestId} = render(`
    <button data-testid="ok-button" type="submit" disabled>
      OK
    </button>
  `)

  expect(queryByTestId('ok-button')).toHaveAttribute('disabled')
  expect(queryByTestId('ok-button')).toHaveAttribute('type')
  expect(queryByTestId('ok-button')).not.toHaveAttribute('class')
  expect(queryByTestId('ok-button')).toHaveAttribute('type', 'submit')
  expect(queryByTestId('ok-button')).not.toHaveAttribute('type', 'button')

  expect(() =>
    expect(queryByTestId('ok-button')).not.toHaveAttribute('disabled'),
  ).toThrowError()
  expect(() =>
    expect(queryByTestId('ok-button')).not.toHaveAttribute('type'),
  ).toThrowError()
  expect(() =>
    expect(queryByTestId('ok-button')).toHaveAttribute('class'),
  ).toThrowError()
  expect(() =>
    expect(queryByTestId('ok-button')).not.toHaveAttribute('type', 'submit'),
  ).toThrowError()
  expect(() =>
    expect(queryByTestId('ok-button')).toHaveAttribute('type', 'button'),
  ).toThrowError()
  expect(() =>
    expect({thisIsNot: 'an html element'}).not.toHaveAttribute(),
  ).toThrowError()
})

test('.toHaveClass', () => {
  const {queryByTestId} = render(`
    <div>
      <button data-testid="delete-button" class="btn extra btn-danger">
        Delete item
      </button>
      <button data-testid="cancel-button">
        Cancel
      </button>
    </div>
  `)

  expect(queryByTestId('delete-button')).toHaveClass('btn')
  expect(queryByTestId('delete-button')).toHaveClass('btn-danger')
  expect(queryByTestId('delete-button')).toHaveClass('extra')
  expect(queryByTestId('delete-button')).not.toHaveClass('xtra')
  expect(queryByTestId('delete-button')).toHaveClass('btn btn-danger')
  expect(queryByTestId('delete-button')).not.toHaveClass('btn-link')
  expect(queryByTestId('cancel-button')).not.toHaveClass('btn-danger')

  expect(() =>
    expect(queryByTestId('delete-button')).not.toHaveClass('btn'),
  ).toThrowError()
  expect(() =>
    expect(queryByTestId('delete-button')).not.toHaveClass('btn-danger'),
  ).toThrowError()
  expect(() =>
    expect(queryByTestId('delete-button')).not.toHaveClass('extra'),
  ).toThrowError()
  expect(() =>
    expect(queryByTestId('delete-button')).toHaveClass('xtra'),
  ).toThrowError()
  expect(() =>
    expect(queryByTestId('delete-button')).not.toHaveClass('btn btn-danger'),
  ).toThrowError()
  expect(() =>
    expect(queryByTestId('delete-button')).toHaveClass('btn-link'),
  ).toThrowError()
  expect(() =>
    expect(queryByTestId('cancel-button')).toHaveClass('btn-danger'),
  ).toThrowError()
})

test('.toHaveStyle', () => {
  const {container} = render(`
    <div class="label" style="background-color: blue; height: 100%">
      Hello World
    </div>
  `)

  const style = document.createElement('style')
  style.innerHTML = `
    .label {
      background-color: black;
      color: white;
      float: left;
    }
  `
  document.body.appendChild(style)
  document.body.appendChild(container)

  expect(container.querySelector('.label')).toHaveStyle(`
    height: 100%;
    color: white;
    background-color: blue;
  `)

  expect(container.querySelector('.label')).toHaveStyle(`
    background-color: blue;
    color: white;
  `)
  expect(container.querySelector('.label')).toHaveStyle(
    'background-color:blue;color:white',
  )

  expect(container.querySelector('.label')).not.toHaveStyle(`
    color: white;
    font-weight: bold;
  `)

  expect(() =>
    expect(container.querySelector('.label')).toHaveStyle('font-weight: bold'),
  ).toThrowError()
  expect(() =>
    expect(container.querySelector('.label')).not.toHaveStyle('color: white'),
  ).toThrowError()

  // Make sure the test fails if the css syntax is not valid
  expect(() =>
    expect(container.querySelector('.label')).not.toHaveStyle(
      'font-weight bold',
    ),
  ).toThrowError()
  expect(() =>
    expect(container.querySelector('.label')).toHaveStyle('color white'),
  ).toThrowError()

  document.body.removeChild(style)
  document.body.removeChild(container)
})

test('.toBeVisible', () => {
  const {container} = render(`
    <div>
      <header>
        <h1 style="display: none">Main title</h1>
        <h2 style="visibility: hidden">Secondary title</h2>
        <h3 style="visibility: collapse">Secondary title</h3>
        <h4 style="opacity: 0">Secondary title</h4>
        <h5 style="opacity: 0.1">Secondary title</h5>
      </header>
      <section style="display: block; visibility: hidden">
        <p>Hello <strong>World</strong></p>
      </section>
    </div>
  `)

  expect(container.querySelector('header')).toBeVisible()
  expect(container.querySelector('h1')).not.toBeVisible()
  expect(container.querySelector('h2')).not.toBeVisible()
  expect(container.querySelector('h3')).not.toBeVisible()
  expect(container.querySelector('h4')).not.toBeVisible()
  expect(container.querySelector('h5')).toBeVisible()
  expect(container.querySelector('strong')).not.toBeVisible()

  expect(() =>
    expect(container.querySelector('header')).not.toBeVisible(),
  ).toThrowError()
  expect(() =>
    expect(container.querySelector('p')).toBeVisible(),
  ).toThrowError()
})
