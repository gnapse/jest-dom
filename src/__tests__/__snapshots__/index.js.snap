// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.toBeInTheDOM 1`] = `
"<dim>expect(</><red>element</><dim>).not.toBeInTheDOM(</><dim>)</>

Received:
  <red><span data-testid=\\"count-value\\" /></>"
`;

exports[`.toBeInTheDOM 2`] = `
"<dim>expect(</><red>element</><dim>).toBeInTheDOM(</><dim>)</>

Received:
  <red>null</>"
`;

exports[`.toBeInTheDOM 3`] = `
"<dim>expect(</><red>received</><dim>).toBeInTheDOM(</><dim>)</>

<red>received</> value must be an HTMLElement.
Received:
  object: <red>{\\"thisIsNot\\": \\"an html element\\"}</>"
`;

exports[`.toBeVisible 1`] = `
"<dim>expect(</><red>element</><dim>).not.toBeVisible(</><dim>)</>

Received element is visible:
  <red><header /></>"
`;

exports[`.toBeVisible 2`] = `
"<dim>expect(</><red>element</><dim>).toBeVisible(</><dim>)</>

Received element is not visible:
  <red><p /></>"
`;

exports[`.toHaveAttribute 1`] = `
"<dim>expect(</><red>element</><dim>).not.toHaveAttribute(</><green><green>\\"disabled\\"<green></><dim>) // element.hasAttribute(\\"disabled\\")</>

Expected the element not to have attribute:
<green>  disabled</>
Received:
<red>  disabled=\\"\\"</>"
`;

exports[`.toHaveAttribute 2`] = `
"<dim>expect(</><red>element</><dim>).not.toHaveAttribute(</><green><green>\\"type\\"<green></><dim>) // element.hasAttribute(\\"type\\")</>

Expected the element not to have attribute:
<green>  type</>
Received:
<red>  type=\\"submit\\"</>"
`;

exports[`.toHaveAttribute 3`] = `
"<dim>expect(</><red>element</><dim>).toHaveAttribute(</><green><green>\\"class\\"<green></><dim>) // element.hasAttribute(\\"class\\")</>

Expected the element to have attribute:
<green>  class</>
Received:
<red>  null</>"
`;

exports[`.toHaveAttribute 4`] = `
"<dim>expect(</><red>element</><dim>).not.toHaveAttribute(</><green><green>\\"type\\"<green></><dim>, </><green><green>\\"submit\\"<green></><dim>) // element.getAttribute(\\"type\\") === \\"submit\\"</>

Expected the element not to have attribute:
<green>  type=\\"submit\\"</>
Received:
<red>  type=\\"submit\\"</>"
`;

exports[`.toHaveAttribute 5`] = `
"<dim>expect(</><red>element</><dim>).toHaveAttribute(</><green><green>\\"type\\"<green></><dim>, </><green><green>\\"button\\"<green></><dim>) // element.getAttribute(\\"type\\") === \\"button\\"</>

Expected the element to have attribute:
<green>  type=\\"button\\"</>
Received:
<red>  type=\\"submit\\"</>"
`;

exports[`.toHaveAttribute 6`] = `
"<dim>expect(</><red>received</><dim>).not.toHaveAttribute(</><dim>)</>

<red>received</> value must be an HTMLElement.
Received:
  object: <red>{\\"thisIsNot\\": \\"an html element\\"}</>"
`;

exports[`.toHaveClass 1`] = `
"<dim>expect(</><red>element</><dim>).not.toHaveClass(</><green><green>\\"btn\\"<green></><dim>)</>

Expected the element not to have class:
<green>  btn</>
Received:
<red>  btn extra btn-danger</>"
`;

exports[`.toHaveClass 2`] = `
"<dim>expect(</><red>element</><dim>).not.toHaveClass(</><green><green>\\"btn-danger\\"<green></><dim>)</>

Expected the element not to have class:
<green>  btn-danger</>
Received:
<red>  btn extra btn-danger</>"
`;

exports[`.toHaveClass 3`] = `
"<dim>expect(</><red>element</><dim>).not.toHaveClass(</><green><green>\\"extra\\"<green></><dim>)</>

Expected the element not to have class:
<green>  extra</>
Received:
<red>  btn extra btn-danger</>"
`;

exports[`.toHaveClass 4`] = `
"<dim>expect(</><red>element</><dim>).toHaveClass(</><green><green>\\"xtra\\"<green></><dim>)</>

Expected the element to have class:
<green>  xtra</>
Received:
<red>  btn extra btn-danger</>"
`;

exports[`.toHaveClass 5`] = `
"<dim>expect(</><red>element</><dim>).not.toHaveClass(</><green><green>\\"btn btn-danger\\"<green></><dim>)</>

Expected the element not to have class:
<green>  btn btn-danger</>
Received:
<red>  btn extra btn-danger</>"
`;

exports[`.toHaveClass 6`] = `
"<dim>expect(</><red>element</><dim>).toHaveClass(</><green><green>\\"btn-link\\"<green></><dim>)</>

Expected the element to have class:
<green>  btn-link</>
Received:
<red>  btn extra btn-danger</>"
`;

exports[`.toHaveClass 7`] = `
"<dim>expect(</><red>element</><dim>).toHaveClass(</><green><green>\\"btn-danger\\"<green></><dim>)</>

Expected the element to have class:
<green>  btn-danger</>
Received:
"
`;

exports[`.toHaveStyle 1`] = `
"<dim>expect(</><red>element</><dim>).toHaveStyle(</><dim>)</>

<green>- Expected</>

<green>- font-weight: bold;</>
<dim>  </>
<red>+ </>"
`;

exports[`.toHaveStyle 2`] = `
"<dim>expect(</><red>element</><dim>).not.toHaveStyle(</><dim>)</>

<dim>Compared values have no visual difference.</>"
`;

exports[`.toHaveStyle 3`] = `"Syntax error parsing expected css: property missing ':' in 1:24"`;

exports[`.toHaveStyle 4`] = `"Syntax error parsing expected css: property missing ':' in 1:18"`;

exports[`.toHaveTextContent 1`] = `
"<dim>expect(</><red>received</><dim>).toHaveTextContent(</><dim>)</>

<red>received</> value must be an HTMLElement.
Received: <red>null</>"
`;

exports[`.toHaveTextContent 2`] = `
"<dim>expect(</><red>element</><dim>).toHaveTextContent(</><dim>)</>

Expected element to have text content:
<green>  3</>
Received:
<red>  2</>"
`;

exports[`.toHaveTextContent 3`] = `
"<dim>expect(</><red>element</><dim>).not.toHaveTextContent(</><dim>)</>

Expected element not to have text content:
<green>  2</>
Received:
<red>  2</>"
`;
